### PR TITLE
feat: add interfaces

### LINKED ISSUE #192: feat: add common interfaces to forge-std
Add ERC-20, 165, 721, 1155, and 4626 interfaces so devs don't need to install OZ or other libs to use them. This matches the set of default interfaces included by vyper: https://github.com/vyperlang/vyper/tree/master/vyper/builtin_interfaces

Note that name and symbol are technically optional in ERC-20, so vyper and OZ both split those into separate files. But I don't think I've seen a token that doesn't have name/symbol, so perhaps we just include that by default in a single IERC20 interface?

h/t @brockelmore

### PR DESCRIPTION
**Pull request by mds1**
Closes #192 

When an EIP had a solidity interface with comments I pasted that directly, otherwise I took the OZ interfaces and truncated them a bit. In all cases I made the natspec style consistent

Includes an unrelated visibility change for bound

Merging into the v0.3 branch since that should get merged soon

### CODE REVIEW

File: src/StdUtils.sol [hunk 1]
```diff
@@ -7,7 +7,7 @@ abstract contract StdUtils {
     uint256 internal constant UINT256_MAX =
         115792089237316195423570985008687907853269984665640564039457584007913129639935;
 
-    function _bound(uint256 x, uint256 min, uint256 max) internal view virtual returns (uint256 result) {
+    function _bound(uint256 x, uint256 min, uint256 max) internal pure virtual returns (uint256 result) {
```
Comment (ZeroEkkusu) [inline src/StdUtils.sol:10]:
```suggestion
    function _bound(uint256 x, uint256 min, uint256 max) private pure virtual returns (uint256 result) {
```

Comment (mds1) [reply src/StdUtils.sol:10]: I figured internal is preferable cause it also enables using `bound` without logging the result

**Review by ZeroEkkusu — APPROVED** [review]

File: src/interfaces/IERC1155.sol [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    event ApprovalForAll(address indexed _owner, address indexed _operator, bool _approved);
+
+    /// @dev MUST emit when the URI is updated for a token ID.
+    /// URIs are defined in RFC 3986.
+    /// The URI MUST point to a JSON file that conforms to the "ERC-1155 Metadata URI JSON Schema".
+    event URI(string _value, uint256 indexed _id);
+
+    /// @notice Transfers `_value` amount of an `_id` from the `_from` address to the `_to` address specified (with safety call).
+    /// @dev Caller must be approved to manage the tokens being transferred out of the `_from` account (see "Approval" section of the standard).
+    /// MUST revert if `_to` is the zero address.
```
[the remark this answers is not in the document]
Comment (mds1) [reply src/interfaces/IERC1155.sol]: Let me know how the fix in 210d516 looks

File: src/interfaces/IERC4626.sol [hunk 3]
```diff
@@ -0,0 +1,189 @@
+pragma solidity >=0.6.2;
+
+import "src/interfaces/IERC20.sol";
+
+/// @dev Interface of the ERC4626 "Tokenized Vault Standard", as defined in
+/// https://eips.ethereum.org/EIPS/eip-4626[ERC-4626].
```
Comment (ZeroEkkusu) [inline src/interfaces/IERC4626.sol:6]:
```suggestion
/// https://eips.ethereum.org/EIPS/eip-4626
```

File: src/interfaces/IERC4626.sol [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    /// NOTE: This calculation MAY NOT reflect the “per-user” price-per-share, and instead should reflect the
+    /// “average-user’s” price-per-share, meaning what the average user should expect to see when exchanging to and
+    /// from.
+    function convertToAssets(uint256 shares) external view returns (uint256 assets);
+
+    /// @notice Returns the maximum amount of the underlying asset that can be deposited into the Vault for the receiver,
+    /// through a deposit call.
+    /// @dev
+    /// - MUST return a limited value if receiver is subject to some deposit limit.
+    /// - MUST return 2///* 256 - 1 if there is no limit on the maximum amount of assets that may be deposited.
```
Comment (ZeroEkkusu) [inline src/interfaces/IERC4626.sol:57]:
```suggestion
    /// - MUST return 2 ** 256 - 1 if there is no limit on the maximum amount of assets that may be deposited.
```

File: src/interfaces/IERC4626.sol [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    /// - MUST revert if all of assets cannot be deposited (due to deposit limit being reached, slippage, the user not
+    ///   approving enough underlying tokens to the Vault contract, etc).
+    ///
+    /// NOTE: most implementations will require pre-approval of the Vault with the Vault’s underlying asset token.
+    function deposit(uint256 assets, address receiver) external returns (uint256 shares);
+
+    /// @notice Returns the maximum amount of the Vault shares that can be minted for the receiver, through a mint call.
+    /// @dev
+    /// - MUST return a limited value if receiver is subject to some mint limit.
+    /// - MUST return 2///* 256 - 1 if there is no limit on the maximum amount of shares that may be minted.
```
Comment (ZeroEkkusu) [inline src/interfaces/IERC4626.sol:90]:
```suggestion
    /// - MUST return 2 ** 256 - 1 if there is no limit on the maximum amount of shares that may be minted.
```

File: src/interfaces/IERC1155.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZeroEkkusu) [reply src/interfaces/IERC1155.sol]: @mds1 Nice, especially with multi-line descriptions. There are more different patterns (e.g. in `IERC721.sol`) but it's not a problem.

[the remark this answers is not in the document]
Comment (mds1) [reply src/interfaces/IERC1155.sol]: Yea, all the ERC's use different styles which made this a bit annoying haha. I think this is probably fine for now as it really only matters for consistent docgen output once `forge doc` is merged, so we can revisit at that time

[the remark this answers is not in the document]
Comment (mds1) [reply src/interfaces/IERC1155.sol]: Feel free to squash merge this into v0.3 once you're done reviewing 🙂

**Pull request merged**